### PR TITLE
Skeletonified client

### LINKED ISSUE #57: Use python3-pip-skeleton
We need to clean up the CI, and drop pipenv in favour of a single pyproject.toml, similar to https://github.com/PandABlocks/PandABlocks-ioc.

We also need to add CI for python 3.10 and 3.11.

### PR DESCRIPTION
**Pull request by evalott100**
Closes #57 

[Skeletonified](https://github.com/DiamondLightSource/python3-pip-skeleton-cli) pandablocks, moved docs to the new structure.

The new docs publishing is working, they can be viewed [here](https://pandablocks.github.io/PandABlocks-client/skeletonified-client/index.html).

### CODE REVIEW

**Review by AlexanderWells-diamond — CHANGES_REQUESTED** [review]
A couple of bits to clean up

File: docs/user/explanations/sans-io.rst [hunk 1]
```diff
@@ -48,7 +48,7 @@ Wrappers
 --------
 
 Of course, these Connections are useless without connecting some I/O. To aid with
-this, wrappers are included for use in `asyncio <library/asyncio>` and blocking programs. They expose
+this, wrappers are included for use in `asyncio <asyncio>` and blocking programs. They expose
```
[the remark this answers is not in the document]
Comment (evalott100) [reply docs/user/explanations/sans-io.rst:51]: Good idea, I can make a new ADR for this and link this document. I'll also make one for using a mock asyncio client in tests.

File: src/pandablocks/__init__.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import sys
+
+if sys.version_info < (3, 8):
+    from importlib_metadata import version  # noqa
```
[the remark this answers is not in the document]
Comment (evalott100) [reply src/pandablocks/__init__.py:4]: I'll have to look into this, codecov does seem to be successfully running 3.7.

image

File: pyproject.toml [hunk 3]
```diff
@@ -1,4 +1,117 @@
...
+license.file = "Apache License 2.0"
+readme = "README.rst"
+requires-python = ">=3.7"
+
+[project.optional-dependencies]
+h5py = ["h5py", "matplotlib"]
+dev = [
+    # A dev install will require [h5py] packages too
+    "h5py",
+    "matplotlib",
```
[the remark this answers is not in the document]
Comment (coretl) [reply pyproject.toml]: This is how ophyd_async does it:
https://github.com/bluesky/ophyd-async/blob/10bfe32d95331833556942c79748a5530d1fddbc/pyproject.toml#L35-L36

File: src/pandablocks/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (evalott100) [reply src/pandablocks/__init__.py:4]: @AlexanderWells-diamond it was just a case of making the test environment 3.7 editable. Now there are two tests of 3.10, one editable, on without.

File: pyproject.toml [hunk 4]
```diff
@@ -6,7 +6,7 @@ build-backend = "setuptools.build_meta"
 description = "A Python client to control and data ports of the PandABlocks TCP server"
 name = "pandablocks"
 classifiers = [
-    "Development Status :: 3 - Alpha",
+    "Development Status :: 3 - Stable",
```
Comment (AlexanderWells-diamond) [inline pyproject.toml:9]: This is not a valid classifier. See here: https://pypi.org/classifiers/

**Review by AlexanderWells-diamond — CHANGES_REQUESTED** [review]
Classifier needs fixing

File: pyproject.toml (hunk 4 above)
Comment (evalott100) [reply pyproject.toml:9]: Ah I suppose
`Development Status :: 5 - Production/Stable`
will do now

**Comment by AlexanderWells-diamond** [comment]
It's not strictly required in this PR, but I'll mention it here: There needs to be a redirect set up from the old documentation site (https://pandablocks-client.readthedocs.io/en/stable/index.html), to the new GitHub pages one. 

This might be doable using [this project](https://github.com/dls-controls/readthedocsredirect), but that was designed to work with DiamondLightSource organization so may require some tweaking. Or doing it manually.

**Comment by evalott100** [comment]
> It's not strictly required in this PR, but I'll mention it here: There needs to be a redirect set up from the old documentation site (https://pandablocks-client.readthedocs.io/en/stable/index.html), to the new GitHub pages one.
> 
> This might be doable using [this project](https://github.com/dls-controls/readthedocsredirect), but that was designed to work with DiamondLightSource organization so may require some tweaking. Or doing it manually.

I can make a ticket for this. What's the purpose of this though - is it just to ensure links to the docs already in place in other projects will redirect to the new link?

**Comment by AlexanderWells-diamond** [comment]
> I can make a ticket for this. What's the purpose of this though - is it just to ensure links to the docs already in place in other projects will redirect to the new link?

Exactly that - we want any links that may be left lying around in odd places, e.g. third party documentation, or wikis, to always go to the right place.

**Comment by evalott100** [comment]
Created a ticket https://github.com/PandABlocks/PandABlocks-client/issues/62

**Review by AlexanderWells-diamond — APPROVED** [review]
Looks good to me.

**Pull request merged**